File: chain-listener/event-register.py
import json
from flask import Flask, request
import os

app = Flask(__name__)

def validate_user_info(info):
    if not info:
        return "Bad Request: event inforamtion", 400
    if not info['chatId']:
        return "Bad Request: missing chat id", 400
    if not info['contractAddress']:
        return "Bad Request: missing contract to monitor", 400
    if not info['eventName']:
        return "Bad Request: missing event to monitor", 400


@app.route('/event-register', methods=['POST'])
def save_user_info():
    user_info = request.get_json()
    # validate_user_info(user_info)

    filename = './chain-listener/event-register.json'

    # Check if the file exists and is not empty
    if os.path.isfile(filename) and os.path.getsize(filename) > 0:
        # Load the existing JSON data
        with open(filename, 'r') as file:
            data = json.load(file)
    else:
        # If the file is empty or doesn't exist, create an empty dictionary
        data = []

    # Append the new data to the existing dictionary
    data.append(user_info)

    # Save the updated dictionary as a JSON file
    with open(filename, 'w') as file:
        json.dump(data, file)
    return "User information saved successfully", 200

if __name__ == '__main__':
    app.run(debug=True, port=3005)

File: event-register.js
const express = require('express');
const fs = require('fs');
const bodyParser = require('body-parser');
const ethers = require('ethers');
const app = express();
app.use(bodyParser.json());

app.post('/event-register', (req, res) => {
    const userInfo = req.body;
    console.log(userInfo);
    const error = validateUserInfo(userInfo);
    if (error) {
        res.status(400).send(error);
        return;
    }

    const filename = './event-register.json';

    let data;
    if (fs.existsSync(filename) && fs.statSync(filename).size > 0) {
        data = JSON.parse(fs.readFileSync(filename, 'utf8'));
    } else {
        data = [];
    }

    // Generate the keccak256 hash of the eventName
    const eventNameHash = ethers.keccak256(ethers.toUtf8Bytes(userInfo['eventName']));
    userInfo['signatureHash'] = eventNameHash;

    data.push(userInfo);
    fs.writeFileSync(filename, JSON.stringify(data));

    res.status(200).send("User information saved successfully");
});

app.listen(3005, () => {
    console.log("Server running on port 3005");
});

function validateUserInfo(info) {
    if (!info) {
        return "Bad Request: event information";
    }
    if (!info['chatId']) {
        return "Bad Request: missing chat id";
    }
    if (!info['contractAddress']) {
        return "Bad Request: missing contract to monitor";
    }
    if (!info['eventName']) {
        return "Bad Request: missing event to monitor";
    }

    return null;
}